### DISCUSSION TITLE
Grooved patterns on a cylinder and/or cone

### DISCUSSION THREAD

**Discussion by shrekuntu** [gumyr/build123d#836] [General]
I need help with creating grooved patterns such as diamond on a cylinder. Basically these are rubber layers let's say 1/2" thick with a grooved diamond patterns 1/8" deep, like one shown on the attachment. To make things harder, these cylinders are most of the time slightly tapered towards the edges with a flat portion in the middle.
![lagging](https://github.com/user-attachments/assets/9a223450-a41c-4158-a762-c86d51228633)

**Comment by gumyr** [comment]
This isn't easy but here is some code to generate the diamond pattern:
```py
from build123d import *
from ocp_vscode import show_all
import copy

radius = 3
pitch = 20
gap = 0.1
repeats = 12

# Create the perimeter Edges of a single diamond
right0 = Pos(Z=-pitch / (2 * repeats) + gap) * Helix(
    pitch=pitch, height=pitch / (2 * repeats) - gap, radius=radius
)
left0 = Pos(Z=-pitch / (2 * repeats) + gap) * Helix(
    pitch=pitch, height=pitch / (2 * repeats) - gap, radius=radius, lefthand=True
)
right1 = right0.mirror(Plane.XY)
left1 = left0.mirror(Plane.XY)

# Create a single diamond
diamond_perimeter = Wire([right0, right1, left0, left1])
diamond = thicken(-Face.make_surface(diamond_perimeter), 0.1)

# Create many diamonds from copies of the single diamond
diamonds = [
    Pos(Z=row * pitch / (2 * repeats))
    * (Rot(Z=(col + (row % 2) * 0.5) * 360 / repeats) * copy.copy(diamond))
    for row in range(3)
    for col in range(repeats)
]

# Create a single diamond pattern object
diamond_pattern = Compound(children=diamonds)

show_all()
```
![image](https://github.com/user-attachments/assets/fd094537-6221-4bfa-a074-5547cc154b56)

The `Helix` class supports a `cone_angle` but that adds significantly to the complexity of the whole thing as each row of diamonds are unique.

**Comment by shrekuntu** [comment]
Thank you @gumyr. I didn't expect it to be easy. I'm just wondering if rather than building diamond tiles, would it be possible to subtract straight line grooves starting with solid plain rubber layer? Before wrapping layer of rubber around the cylinder and gluing it
, it's a straight line groove on a flat rubber sheet
![chevron-lagging](https://github.com/user-attachments/assets/ea74db64-9170-4374-8d06-b956d8bcbbf1)

**Reply by gumyr** [reply]
We're working within the limitations of the OpenCascade CAD core here; unfortunately, subtractions like that tend to be problematic and slow. 

However, this roller is easier to build as the grooves only go one way:
```py
radius = 3
pitch = 20
gap = 0.1
repeats = 24

with BuildSketch() as profile:
    Circle(radius)
    Circle(radius - 2 * gap, mode=Mode.SUBTRACT)
    with PolarLocations(radius, repeats):
        Rectangle(gap, gap, mode=Mode.SUBTRACT)

roller = Solid.extrude_linear_with_rotation(
    profile.face(), center=(0, 0, 0), normal=(0, 0, pitch), angle=90
)
```
![image](https://github.com/user-attachments/assets/75b038b0-6f0b-441d-afd5-0901743980e9)

By the way - if one tries to intersect this roller will one with the opposite twist angle to create the diamond it takes a long time (beyond my patience - maybe it would work).

**Reply by shrekuntu** [reply]
Thank you very much @gumyr. This will be a great start, your solution looks elegant with just a few lines of code. I think I can take it from here.
In this case of final assembly, my users will not wait in real time, I can push this through a background job processing, it will be fine.

**Comment by shrekuntu** [comment]
Sorry, I spoke too soon. When I try to run your code, I get the following error related to Solid.extrude_linear_with_rotation():
```
TypeError: Add(): incompatible function arguments. The following argument types are supported:
    1. (self: OCP.TopoDS.TopoDS_Builder, S: OCP.TopoDS.TopoDS_Shape, C: OCP.TopoDS.TopoDS_Shape) -> None

Invoked with: <OCP.TopoDS.TopoDS_Builder object at 0x7f48dc78b0b0>, <OCP.TopoDS.TopoDS_Compound object at 0x7f48e55b0e70>, <build123d.topology.Shape object at 0x7f48dc7f7390>

```
I even upgraded to latest CQEditor jgenestein fork, still the same error.

**Reply by gumyr** [reply]
In this statement:
```py
with BuildSketch() as profile:
```
`profile` refers to the **Builder** not the sketch within.  I used `profile.face()` to extract the face from the Sketch, did you do that?

**Reply by shrekuntu** [reply]
I ran your code as is, just added build123d import at the top. 

```
from build123d import *

radius = 3
pitch = 20
gap = 0.1
repeats = 24

with BuildSketch() as profile:
    Circle(radius)
    Circle(radius - 2 * gap, mode=Mode.SUBTRACT)
    with PolarLocations(radius, repeats):
        Rectangle(gap, gap, mode=Mode.SUBTRACT)

roller = Solid.extrude_linear_with_rotation(
    profile.face(), center=(0, 0, 0), normal=(0, 0, pitch), angle=90
)
```

**Reply by gumyr** [reply]
Sorry, I replied in haste.  I'm working in another branch which doesn't seem to have this problem.  It will take some investigation to see what is going on.

**Reply by shrekuntu** [reply]
No problem. Take your time. Thank you @gumyr.
